src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = Object.keys(currentState).reduce((newState, key) => {
+          if (!action.keysToRemove.includes(key)) {
+            newState[key] = currentState[key];
+          }
+
+          return newState;
+        }, {});
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+    stateHistory.push({ ...currentState });
```

Comment:
The variable name 'currentState' is not descriptive enough about the fact that it's a copy of the state that will be modified. A better name would be 'stateCopy' or 'nextState' to convey its purpose more clearly.